bridge/nimbus/public_whisper_api.go

```diff
@@ -0,0 +1,192 @@
+		keyC := C.CString(key)
+		defer C.free(unsafe.Pointer(keyC))
+
+		c <- C.nimbus_delete_keypair(keyC)
+	})
+	if err, ok := retVal.(error); ok {
+		return false, err
+	}
+
+	return retVal.(bool), nil
+}
+
+// NewMessageFilter creates a new filter that can be used to poll for
+// (new) messages that satisfy the given criteria.
+func (w *nimbusPublicWhisperAPIWrapper) NewMessageFilter(req whispertypes.Criteria) (string, error) {
+	// topics := make([]whisper.TopicType, len(req.Topics))
+	// for index, tt := range req.Topics {
+	// 	topics[index] = whisper.TopicType(tt)
+	// }
+
+	// criteria := whisper.Criteria{
+	// 	SymKeyID:     req.SymKeyID,
+	// 	PrivateKeyID: req.PrivateKeyID,
+	// 	Sig:          req.Sig,
+	// 	MinPow:       req.MinPow,
+	// 	Topics:       topics,
+	// 	AllowP2P:     req.AllowP2P,
+	// }
+	// return w.publicWhisperAPI.NewMessageFilter(criteria)
+	// TODO
+	return "", errors.New("not implemented")
+}
+
+// GetFilterMessages returns the messages that match the filter criteria and
+// are received between the last poll and now.
+func (w *nimbusPublicWhisperAPIWrapper) GetFilterMessages(id string) ([]*whispertypes.Message, error) {
+	idC := C.CString(id)
+	defer C.free(unsafe.Pointer(idC))
+
+	var (
+		messageList *list.List
+		ok          bool
+	)
+	if messageList, ok = (*w.filterMessages)[id]; !ok {
+		return nil, fmt.Errorf("no filter with ID %s", id)
+	}
+
+	retVal := make([]*whispertypes.Message, messageList.Len())
+	if messageList.Len() == 0 {
+		return retVal, nil
+	}
+
+	// TODO: make this thread-safe
+	elem := messageList.Front()
+	index := 0
+	for elem != nil {
+		retVal[index] = (elem.Value).(*whispertypes.Message)
+		index++
+		next := elem.Next()
+		messageList.Remove(elem)
+		elem = next
+	}
+	return retVal, nil
+}
+
+// Post posts a message on the Whisper network.
+// returns the hash of the message in case of success.
+func (w *nimbusPublicWhisperAPIWrapper) Post(ctx context.Context, req whispertypes.NewMessage) ([]byte, error) {
+	retVal := w.routineQueue.Send(func(c chan<- interface{}) {
+		msg := C.post_message{
+			ttl:       C.uint(req.TTL),
+			powTime:   C.double(req.PowTime),
+			powTarget: C.double(req.PowTarget),
+		}
+		if req.SigID != "" {
+			msg.sourceID = C.CString(req.SigID)
+			defer C.free(unsafe.Pointer(msg.sourceID))
+		}
+		if req.SymKeyID != "" {
+			msg.symKeyID = C.CString(req.SymKeyID)
+			defer C.free(unsafe.Pointer(msg.symKeyID))
+		}
+		if req.PublicKey != nil && len(req.PublicKey) > 0 {
+			msg.pubKey = (*C.uchar)(C.CBytes(req.PublicKey))
+			defer C.free(unsafe.Pointer(msg.pubKey))
+		}
+		msg.payloadLen = C.size_t(len(req.Payload))
+		msg.payload = (*C.uchar)(C.CBytes(req.Payload))
+		defer C.free(unsafe.Pointer(msg.payload))
+		msg.paddingLen = C.size_t(len(req.Padding))
+		msg.padding = (*C.uchar)(C.CBytes(req.Padding))
+		defer C.free(unsafe.Pointer(msg.padding))
+		copyTopicToCBuffer(&msg.topic[0], req.Topic[:])
+
+		if C.nimbus_post(&msg) {
+			c <- make([]byte, 0)
```

Comment:
Why do we send an empty slice?

Reply:
This is temporary code until Nimbus starts returning envelope hashes on success. I'll ping @kdeme to see if we can get that in Nimbus, this kind of slipped through the cracks as it is not required right now.

Reply:
Kim said he'll implement this in Nimbus in the afternoon, and then I'll be able to make the necessary changes on the Go side.

Reply:
I've discussed with Adam and we agreed to leave this for a subsequent PR, since the envelope hash isn't being used anywhere for the Nimbus implementation.